proof/access-control/ARM/ArchAccess.thy

```diff
@@ -0,0 +1,287 @@
+  "vspace_cap_rights_to_auth r \<equiv>
+     (if AllowWrite \<in> r then {Write} else {})
+   \<union> (if AllowRead \<in> r then {Read} else {})"
+
+definition arch_cap_auth_conferred where
+  "arch_cap_auth_conferred arch_cap \<equiv>
+     (if is_page_cap arch_cap then vspace_cap_rights_to_auth (acap_rights arch_cap) else {Control})"
+
+subsection \<open>Generating a policy from the current ASID distribution\<close>
+
+definition pte_ref where
+  "pte_ref pte \<equiv> case pte of
+     SmallPagePTE addr atts rights
+       \<Rightarrow> Some (ptrFromPAddr addr, pageBitsForSize ARMSmallPage, vspace_cap_rights_to_auth rights)
+   | LargePagePTE addr atts rights
+       \<Rightarrow> Some (ptrFromPAddr addr, pageBitsForSize ARMLargePage, vspace_cap_rights_to_auth rights)
+   | _ \<Rightarrow> None"
+
+definition pde_ref2 where
+  "pde_ref2 pde \<equiv> case pde of
+     SectionPDE addr atts domain rights
+       \<Rightarrow> Some (ptrFromPAddr addr, pageBitsForSize ARMSection, vspace_cap_rights_to_auth rights)
+   | SuperSectionPDE addr atts rights
+       \<Rightarrow> Some (ptrFromPAddr addr, pageBitsForSize ARMSuperSection, vspace_cap_rights_to_auth rights)
+   | PageTablePDE addr atts domain
+       \<comment> \<open>The 0 is a hack, saying that we own only addr, although 12 would also be OK\<close>
+       \<Rightarrow> Some (ptrFromPAddr addr, 0, {Control})
+   | _ \<Rightarrow> None"
+
+\<comment> \<open>We exclude the global page tables from the authority graph. Alternatively, we could include them
+    and add a wellformedness constraint to policies that requires that every label has the necessary
+    authority to whichever label owns the global page tables, so that when a new page directory is
+    created and references to the global page tables are added to it, no new authority is gained.
+    Note: excluding the references to global page tables in this way brings in some ARM
+          arch-specific VM knowledge here\<close>
+definition vs_refs_no_global_pts :: "kernel_object \<Rightarrow> (obj_ref \<times> obj_ref \<times> aa_type \<times> auth) set"
+  where
+  "vs_refs_no_global_pts \<equiv> \<lambda>ko. case ko of
+     ArchObj (ASIDPool pool) \<Rightarrow> (\<lambda>(r,p). (p, ucast r, AASIDPool, Control)) ` graph_of pool
+   | ArchObj (PageDirectory pd) \<Rightarrow>
+       \<Union>(r,(p, sz, auth)) \<in> graph_of (pde_ref2 o pd) - {(x,y). (ucast (kernel_base >> 20) \<le> x)}.
+         (\<lambda>(p, a). (p, ucast r, APageDirectory, a)) ` (ptr_range p sz \<times> auth)
+   | ArchObj (PageTable pt) \<Rightarrow>
+       \<Union>(r,(p, sz, auth)) \<in> graph_of (pte_ref o pt).
+         (\<lambda>(p, a). (p, ucast r, APageTable, a)) ` (ptr_range p sz \<times> auth)
+   | _ \<Rightarrow> {}"
+
+definition state_vrefs where
+  "state_vrefs s = case_option {} vs_refs_no_global_pts o kheap s"
+
+end
+
+
+context pspace_update_eq begin interpretation Arch .
```

Comment:
Yes, this just says that if we have some function `f` that preserves the heap (`pspace_update_eq`), then we expect it to preserve `state_vrefs`, but that we temporarily need access to `Arch` to prove that. Using `interpretation` instead of `sublocale` ensures that the access is only temporary.